ariadne/tracknet_v2_1/processor.py

```diff
@@ -81,72 +76,77 @@ def preprocess_chunk(self,
                          idx: str) -> ProcessedTracknetDataChunk:
         chunk_df = chunk.df_chunk_data
         chunk_id = int(chunk_df.event.values[0])
-        output_name = (self.output_dir + '/tracknet_%s_%d' % (idx, chunk_id))
+        output_name = os.path.join(self.output_dir, f'tracknet_{idx.replace(".txt", "")}')
         self.chunks.append(chunk_id)
-
         return ProcessedTracknetDataChunk(chunk_df, output_name, chunk_id)
 
-    def cartesian(self, df1, df2):
-        rows = itertools.product(df1.iterrows(), df2.iterrows())
-        df = pd.DataFrame(left.append(right) for (_, left), (_, right) in rows)
-        df_fakes = df[(df['track_left'] == -1) & (df['track_right'] == -1)]
-        df = df[(df['track_left'] != df['track_right'])]
-        df = pd.concat([df, df_fakes], axis=0)
-        df = df.sample(frac=1).reset_index(drop=True)
-        return df.reset_index(drop=True)
-
     def postprocess_chunks(self,
                            chunks: List[ProcessedTracknetDataChunk]) -> ProcessedTracknetData:
         for chunk in chunks:
+            if chunk.processed_object is None or len(chunk.processed_object)==0:
+                continue
             chunk_data_x = []
             chunk_data_y = []
             chunk_data_len = []
             chunk_data_real = []
             chunk_data_moment = []
             df = chunk.processed_object
+
             grouped_df = df[df['track'] != -1].groupby('track')
+            last_station = df[df['station'] > 1][['phi', 'z']].values
+            if grouped_df.ngroups == 0:
+                print('multiplicity is 0!')
+                chunk.processed_object = None
+                continue
             for i, data in grouped_df:
                 chunk_data_x.append(data[['r', 'phi', 'z']].values[:-1])
-                chunk_data_y.append(data[['r', 'phi', 'z']].values[-1])
+                chunk_data_y.append(data[['phi', 'z']].values[-1])
                 chunk_data_len.append(2)
                 chunk_data_moment.append(data[['px','py','pz']].values[-1])
                 chunk_data_real.append(1)
             print('=====> id', chunk.id)
+            multiplicity = len(chunk_data_x)
             first_station = df[df['station'] == 0][['r', 'phi', 'z', 'track']]
             first_station.columns = ['r_left', 'phi_left', 'z_left', 'track_left']
 
             second_station = df[df['station'] == 1][['r', 'phi', 'z', 'track']]
             second_station.columns = ['r_right', 'phi_right', 'z_right', 'track_right']
 
-            fake_tracks = self.cartesian(first_station, second_station)
-            fake_tracks = fake_tracks.sample(n=int(fake_tracks.shape[0] / 10), random_state=1)
+            fake_tracks = cartesian(first_station, second_station)
+            fake_tracks = fake_tracks.sample(n=int(fake_tracks.shape[0] / 2), random_state=1)
```

Comment:
`n=len(fake_tracks) // 2` or `n=fake_tracks.shape[0] // 2`

Reply:
why only a half?

Reply:
removed this